common/src/collision.rs

```diff
@@ -0,0 +1,477 @@
+use crate::node::DualGraph;
+use crate::{dodeca::Vertex, graph::NodeId};
+
+#[allow(dead_code)]
+/*
+The set of voxels that a collision body covers within a chunk
+*/
+#[derive(PartialEq, Clone, Debug, Copy)]
+pub struct ChunkBoundingBox {
+    pub node: NodeId,
+    pub chunk: Vertex,
+    pub min_xyz: na::Vector3<u32>,
+    pub max_xyz: na::Vector3<u32>,
+    pub dimension: u8,
+}
+
+pub struct VoxelAddress {
+    pub node: NodeId,
+    pub chunk: Vertex,
+    pub index: u32,
+}
+
+/*
+The set of voxels that a collision body covers.
+*/
+pub struct BoundingBox {
+    pub bounding_boxes: Vec<ChunkBoundingBox>,
+}
+
+// from a node coordinate in an arbitrary node, returns the which chunk the point would reside in
+pub fn chunk_from_location(location: na::Vector4<f64>) -> Option<Vertex> {
+    for v in Vertex::iter() {
+        let pos = (v.chunk_to_node().try_inverse().unwrap() * location).xyz();
+        if (pos.x >= 0_f64)
+            && (pos.x <= 1_f64)
+            && (pos.y >= 0_f64)
+            && (pos.y <= 1_f64)
+            && (pos.z >= 0_f64)
+            && (pos.z <= 1_f64)
+        {
+            return Some(v);
+        }
+    }
+    None
+}
+
+// Does not figure out which chunk it is in automatically
+// The pattern I use to unwrap the results of neighbor is kind of awkward.
+impl BoundingBox {
+    pub fn create_aabb(
+        start_node: NodeId,
+        position: na::Vector4<f64>,
+        radius: f64,
+        graph: &DualGraph,
+        dimension: u8,
+    ) -> Self {
+        assert!(
+            radius <= 1.0,
+            "Error: the radius of a bounding box may not exceed 1 absolute unit."
+        );
+
+        let start_chunk =
+            chunk_from_location(position).expect("Error: cannot find chunk for given position.");
+        let mut bounding_boxes = Vec::<ChunkBoundingBox>::new();
+        let sides = start_chunk.canonical_sides();
+
+        // get BBs for the chunks within the node.
+        for v in Vertex::iter() {
+            Self::add_sub_bb(
```

Comment:
If the player is close to the center of a chunk edge, I believe this may fail to find a collision. Much like there are five foreign chunks sharing a dodeca side that need to be checked, there should be two foreign chunks sharing a dodeca edge.